packages/vm/src/mock/storage.rs

```diff
@@ -0,0 +1,273 @@
+use std::collections::BTreeMap;
+#[cfg(feature = "iterator")]
+use std::{
+    iter,
+    ops::{Bound, RangeBounds},
+};
+
+use crate::{FfiResult, ReadonlyStorage, Storage};
+#[cfg(feature = "iterator")]
+use cosmwasm_std::{Order, KV};
+
+#[derive(Default)]
+pub struct MemoryStorage {
+    data: BTreeMap<Vec<u8>, Vec<u8>>,
+}
+
+impl MemoryStorage {
+    pub fn new() -> Self {
+        MemoryStorage::default()
+    }
+}
+
+impl ReadonlyStorage for MemoryStorage {
+    fn get(&self, key: &[u8]) -> FfiResult<Option<Vec<u8>>> {
+        Ok(self.data.get(key).cloned())
+    }
+
+    #[cfg(feature = "iterator")]
+    /// range allows iteration over a set of keys, either forwards or backwards
+    /// uses standard rust range notation, and eg db.range(b"foo"..b"bar") also works reverse
+    fn range<'a>(
+        &'a self,
+        start: Option<&[u8]>,
+        end: Option<&[u8]>,
+        order: Order,
+    ) -> FfiResult<Box<dyn Iterator<Item = FfiResult<KV>> + 'a>> {
```

Comment:
Okay, so copying all this code is to give us the proper Result/Error type for the iterator (and other ops). It seems necessary to match the different function signatures, but it is also missing any code that would let use test the error cases. I would propose replacing this with the following:

1. A wrapper that takes `cosmwasm_std::Storage` and converts to `cosmwasm_vm::Storage` (different return types). This would convert any `StdError` to `FfiError::Other` by stringifying it. It lets us easily reuse code for the success-case.
2. An implementation that always returns an error, where you can set an `FfiError` to in the struct, and all calls will return this error (cloned). I think it is critical to test this error handling - which is a large part of the refactor

Reply:
Another case for this is the fact that staking contract is failing.

https://app.circleci.com/pipelines/github/CosmWasm/cosmwasm/1049/workflows/6581f6b9-4460-42d5-83b6-3467606c2680/jobs/7649/steps

My integration test customizes the Querier (staking module), but that uses the cosmwasm_std::Extern. Having no (test-safe) translation to cosmwasm_vm::Extern makes life hard here (and a whole lot more code copying if I want to support this). I think some transforms allow the two to diverge, but let us easily re-use test helpers. Such transforms can be in the `testing`/`mock` modules to signify they should only be used for test code (like `MockStorage`, `MockQuerier`, etc)

Reply:
Uhm, i just fixed the staking contract integration tests by changing an import and tweaking a function signature to reflect a change that was made after i forked. i just pushed it but everything's passing on my machine, so it should pass here in a few minutes as well.
As for the suggestion in your first comment on this thread, that sounds good, but i don't think there's a strong need to do this right now. there were already tests that checked for errors in a lot of the code i touched, and i had to modify those tests to reflect my changes. Most of what i did is already backed by tests. The only thing that i feel that i should write a test for, is the case where things run out of gas, but i don't want to do it before i've completed the full error flow for that scenario.

Reply:
Yes, these were ideas for future enhancements rather than in this pr.